### PR TITLE
My Jetpack: Add license key selector in License Activation screen.

### LINKED ISSUE #22913: Activate Product key: help navigating Licensing options
### Impacted plugin

Jetpack

### Steps to Reproduce

Currently, when I create a new jetpack site I end up with this notice. 
Screen Shot 2022-02-15 at 3 30 29 PM

Which leads me to 2 screens. 

First link Activate it now. 

Screen Shot 2022-02-15 at 3 34 50 PM

or 

WordPress_com

### A clear and concise description of what you expected to happen.

I expected to get a bit more help with the context here. 
1. You should be able to tell me more about what Produce I have the key for. 

2. On the activation screen. Would it be possible to have a drop-down of the different possible keys? 

3. On the Purchases screen can we highlight the pending activation products keys that are there with an easy way to copy the key. 

### What actually happened

It is really hard to know how to find the key that I need to use to activate the product. 
There is a lot of missing contexts here. 

Granted I didn't just buy a key. I bought the key a long time ago. 

### Other information

_No response_

### Operating System

macOS

### OS Version

_No response_

### Browser

Firefox

### Browser Version(s)

_No response_

### PR DESCRIPTION
**Pull request by jkguidaven**
Fixes #22913

Currently, the Activation UI shows an input box where users have to manually type a license key to activate a product, as shown in the attached capture.

We would like to replace the input box with a dropdown that list the user's available license keys and supports manually entering a key also.

**Before**
Screen Shot 2022-11-25 at 9 41 23 PM

**After**
Screen Shot 2022-11-25 at 9 42 13 PM

#### Changes proposed in this Pull Request:

* Replace the input box on the Activation UI with a dropdown that shows the user's available licenses and supports entering a license key manually

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
Context: 1202858161751496-as-1201858908716176

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
1. Launch a test site that uses a build of the Jetpack plugin based on this branch
2. Connect your site to Jetpack, if it's not already
3. Make sure you do not have a jetpack license key that is not yet activated
4. Go to My Jetpack > Scroll down and click Activate a license
Screen Shot 2022-11-25 at 9 48 39 PM
5. Confirm that there is no license key selector rendered (only the input box).
Screen Shot 2022-11-25 at 9 41 23 PM

6. Open another tab go to https://cloud.jetpack.com/pricing and purchase any Jetpack product (make sure you do not activate it to any site)
7. Go back to the license key activation screen and refresh the page
8. confirm that the screen renders a license key selection.
Screen Shot 2022-11-25 at 9 42 13 PM
9. select the new license key you purchase and activate. Confirm that the license key **successfully activated**.

#### Testing manual input
1. Purchase another product in the pricing page https://cloud.jetpack.com/pricing
2. Go back to the license key activation screen and select 'I want to add a license key manually'.
204000743-94cc9754-6a4b-4f6e-9b03-1dd74751c28d

3. Confirm that the license key **input box** is rendered.
Screen Shot 2022-12-06 at 2 37 52 PM

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - 0-as-1201858908716176

### CODE REVIEW

File: projects/js-packages/licensing/components/activation-screen-controls/style.scss
Comment (jkguidaven) [inline projects/js-packages/licensing/components/activation-screen-controls/style.scss]: I have trouble overriding the label's font styling in the Select component. the `!important` simplifies the fix. Though not desirable but this should be safe solution.

File: projects/js-packages/licensing/components/activation-screen/index.jsx
Comment (manzoorwanijk) [inline projects/js-packages/licensing/components/activation-screen/index.jsx]: I see this error here because the REST API seems to return an object with numeric indices instead of an array. Little bit of debugging tells me that it's something that comes from WPCOM API inside `projects/packages/licensing/src/class-endpoints.php` in `Endpoints::get_user_licenses()`

Screenshot 2022-11-28 at 12 02 55 PM

Comment (manzoorwanijk) [reply projects/js-packages/licensing/components/activation-screen/index.jsx]: Created a patch to fix the bug on the backend D93569-code

**Review by manzoorwanijk — COMMENTED** [review]
Can we ensure that it matches the design given here 0-as-1201858908716176/1201862178408950/f ?

Screenshot 2022-11-28 at 4 04 51 PM

File: projects/js-packages/licensing/components/activation-screen/index.jsx
Comment (jkguidaven) [reply projects/js-packages/licensing/components/activation-screen/index.jsx]: thanks for the quick fix @manzoorwanijk 👍

**Comment by jkguidaven** [comment]
> Can we ensure that it matches the design given here 0-as-1201858908716176/1201862178408950/f ?
> 
> Screenshot 2022-11-28 at 4 04 51 PM

Hi @manzoorwanijk I was hoping to tackle this in a separate PR so to limit the changes.

**Comment by manzoorwanijk** [comment]
> I was hoping to tackle this in a separate PR so to limit the changes.

Sounds good to me @jkguidaven

**Comment by keoshi** [comment]
I couldn't test this despite having several unactivated licenses in my account:

image

You can see a video walkthrough of my experience here and the behavior where the UI is “Fetching available licenses” but then turns into a free text input.

https://user-images.githubusercontent.com/390760/204287224-ff757be0-ce72-4a9f-a383-7945eebe15ef.mov

Also related though outside the scope of this PR: when on the license activation page, no JITMs should be displayed here:

image

File: projects/js-packages/licensing/components/activation-screen/index.jsx
Comment (manzoorwanijk) [inline projects/js-packages/licensing/components/activation-screen/index.jsx]: Instead of calling REST API directly here, I would suggest making use of fetching the data via the redux store so that the data is reusable elsewhere.

Here is what I would suggest doing:
- Rename `updateUserLicenses` to `fetchUserLicenses` in `projects/plugins/jetpack/_inc/client/state/licensing/actions.js`
- Dispatch the above `fetchUserLicenses()` action here inside `useEffect` or use a Query component
- Rename `getDetachedLicensesLoadingInfo` selector to `getIsLoadingLicenses` in `projects/plugins/jetpack/_inc/client/state/licensing/reducer.js`
- Use `getIsLoadingLicenses` selector in `projects/js-packages/licensing/components/activation-screen/index.jsx` to get the loading state.
- Use `getDetachedLicenses` selector to get the licenses.

**Review by manzoorwanijk — CHANGES_REQUESTED** [review]
I would suggest to do it the redux way and to reuse the existing code as much as possible.

**Comment by manzoorwanijk** [comment]
> I couldn't test this despite having several unactivated licenses in my account:

@keoshi Can you please check your browser console to check if the error is the same as mine here https://github.com/Automattic/jetpack/pull/27609#discussion_r1033167198

If yes, can you please try again? I already pushed a fix for that a little while ago via D93569-code

**Comment by keoshi** [comment]
@manzoorwanijk Yes, it works now! Thanks for that.

File: projects/js-packages/licensing/components/activation-screen/index.jsx
Comment (jkguidaven) [reply projects/js-packages/licensing/components/activation-screen/index.jsx]: @manzoorwanijk, I initially thought of using redux to implement this new feature but I realized there is more things to setup on this package before I can do it in redux way. I actually noticed one package that has [reusable code](https://github.com/Automattic/jetpack/tree/trunk/projects/plugins/jetpack/_inc/client/state/licensing) I can probably reuse but I am not sure how to integrate it in the licensing package or is it even good idea to do it since it was structured  in a way it is only reusable within Jetpack plugin.

But now there is strong opinion to just implement it w/ redux I'll try and figure out how to reuse the code.

**Comment by keoshi** [comment]
Everything seems to be working now and it feels great to finally have this in Jetpack!

The only two things I caught UX-wise are:
1. The field changes from a dropdown selector to a free text input when the last option “I want to add a license key manually” is selected.
2. When this option is selected and no text has been entered, the primary button is enabled.

The follow video covers both things:

https://user-images.githubusercontent.com/390760/204294515-7aea8f9a-a441-4ef5-a29b-2eb5203893aa.mov

To solve these issues I'd suggest:
- Showing an additional free text input when that last option is selected where a user could enter a license key manually. This should give everyone the flexibility to go back and activate any of the listed licenses without having to reload the page.
- Disabled the button when this last option is selected AND no text has been entered in the free text input.

These shouldn't block the PR, I don't think, and are things we can address in a follow up PR. cc @ederrengifo

File: projects/js-packages/licensing/components/activation-screen-controls/style.scss
Comment (mattgawarecki) [reply projects/js-packages/licensing/components/activation-screen-controls/style.scss]: I _highly_ recommend against using `!important`, and if I'm not mistaken, it's all but forbidden in our coding conventions. Instead of this approach, add extra specifiers to the rule until it "sticks." For example, we can do multiple classes, or use more parent or child elements, or even add an ID if it's feasible.

Comment (mattgawarecki) [reply projects/js-packages/licensing/components/activation-screen-controls/style.scss]: @jkguidaven After some experimentation, I found that using `.components-input-control__label` instead of `label` gives the necessary specificity to override the rule causing the issue. That is:

```css
  .jp-license-activation-screen-controls--license-field,
  .jp-license-activation-screen-controls--license-field-with-error {
    .components-input-control__label {
```

**Comment by jkguidaven** [comment]
> Everything seems to be working now and it feels great to finally have this in Jetpack!
> 
> The only two things I caught UX-wise are:
> 
> 1. The field changes from a dropdown selector to a free text input when the last option “I want to add a license key manually” is selected.
> 2. When this option is selected and no text has been entered, the primary button is enabled.
> 
> The follow video covers both things:
> 
>  Screen.Recording.2022-11-28.at.13.47.42.mov 
> To solve these issues I'd suggest:
> 
> * Showing an additional free text input when that last option is selected where a user could enter a license key manually. This should give everyone the flexibility to go back and activate any of the listed licenses without having to reload the page.
> * Disabled the button when this last option is selected AND no text has been entered in the free text input.
> 
> These shouldn't block the PR, I don't think, and are things we can address in a follow up PR. cc @ederrengifo

@keoshi thanks for the valuable UX feedbacks and suggestions. 👍 I agree having the input box rendered together w/ the dropdown selector provides more flexibility to the user. I'll try to see if I can have it in this PR (or a follow up PR).

File: projects/js-packages/licensing/components/activation-screen-controls/style.scss
Comment (jkguidaven) [reply projects/js-packages/licensing/components/activation-screen-controls/style.scss]: @mattgawarecki I actually used `.components-input-control__label` before and give up. For some reason the styling doesn't get applied due to some arbitrary class name.. 🤔 I am not sure what I am missing. I used this `jetpack build packages/my-jetpack` command to build it. 

Screen Shot 2022-11-29 at 2 34 05 PM


This arbitrary class name (.css-1imalal) is added to the label which overrides my styling. I can't figure out how I can prevent this from happening aside using `!important`.

Screen Shot 2022-11-29 at 2 39 39 PM

Comment (mattgawarecki) [reply projects/js-packages/licensing/components/activation-screen-controls/style.scss]: I went back to test again and I see what you mean. Another alternative solution here is to apply the same class name multiple times (which is [what's happening inside the WordPress label component](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/input-control/styles/input-control-styles.tsx#L208) to apply `css-1imalal` three times). For example:

```sass
.jp-license-activation-screen-controls--license-field,
.jp-license-activation-screen-controls--license-field-with-error {
    .components-input-control__label.components-input-control__label.components-input-control__label {
```

I'm not sure this is really an improvement over `!important`, but it guards against future usages of `!important`, which can really make styling code hard to work with in a codebase as large as Jetpack. I leave the final decision up to you. 🙂

Comment (jkguidaven) [reply projects/js-packages/licensing/components/activation-screen-controls/style.scss]: thanks @mattgawarecki, I didn't know this approach and it works 👍

**Comment by bindlegirl** [comment]
@jkguidaven is this PR ready for the Crew review or do you still need to address any comments? If yes, could you please check out these conflicts first?

**Comment by jkguidaven** [comment]
> @jkguidaven is this PR ready for the Crew review or do you still need to address any comments? If yes, could you please check out these conflicts first?

Hi @bindlegirl 👋 , I still have some comments to address before it is ready for crew review.

**Comment by bindlegirl** [comment]
@jkguidaven No problem. When ready, just change the label to Needs Review and we'll check it out 👍

**Comment by jkguidaven** [comment]
@manzoorwanijk I have updated the My Jetpack package to load a redux state of the current user licenses which is fed to the `ActivationScreen` component. This should allow future components to utilized detached license key data within My Jetpack component.  It doubled the file changes pretty quick and triggered version bumps 😆 

can you do a final code review?

**Comment by jkguidaven** [comment]
@keoshi I have updated my PR to include your UX suggestions. can you do a final review on this PR?

**Comment by manzoorwanijk** [comment]
@jkguidaven I think we will need to avoid using the word detached licenses because, after D93571-code, we will allow users to transfer their licenses between sites. So, we do not want only to show detached licenses here.
It may be better if the dropdown shows the site that a license is currently attached to, but that may be beyond the scope of this PR.

**Comment by jkguidaven** [comment]
> @jkguidaven I think we will need to avoid using the word detached licenses because, after D93571-code, we will allow users to transfer their licenses between sites. So, we do not want only to show detached licenses here. It may be better if the dropdown shows the site that a license is currently attached to, but that may be beyond the scope of this PR.

Do you have idea when that diff gonna be merge? I'll probably hold merging this PR until that one gets merged.

For the meantime, I'll update the context and use 'Available licenses'.

**Comment by manzoorwanijk** [comment]
> Do you have idea when that diff gonna be merge? I'll probably hold merging this PR until that one gets merged.

That should land soon.

**Comment by manzoorwanijk** [comment]
@jkguidaven D93571-code is deployed now.

**Comment by jkguidaven** [comment]
> @jkguidaven D93571-code is deployed now.

@manzoorwanijk since the current license query we used is only limited to the current site and not for all the licenses from other sites. I would leave the filter for detached license as is.  

For now, since we don't have yet the proper dropdown component that can display more license information in a presentable way (Like the one in the design). We will stick with the current query.

In case user wants to use license from different site, the user can opt for the manual input.

**Comment by manzoorwanijk** [comment]
> > @jkguidaven D93571-code is deployed now.
> 
> @manzoorwanijk since the current license query we used is only limited to the current site and not for all the licenses from other sites. I would leave the filter for detached license as is.
> 
> For now, since we don't have yet the proper dropdown component that can display more license information in a presentable way (Like the one in the design). We will stick with the current query.
> 
> In case user wants to use license from different site, the user can opt for the manual input.

Alright, thank you for the input.

**Review by manzoorwanijk — DISMISSED** [review]
Looks good to me now

**Comment by sdixon194** [comment]
This works for me! Something I noted that might be good for a follow up, I still get the JITM that takes me to: https://causal-catfish.jurassic.ninja/wp-admin/admin.php?page=jetpack#/license/activation

Screen Shot 2022-12-12 at 10 01 15 AM

That license activation page seems different and doesn't let me pick existing licenses from a dropdown. Not sure if there are plans to cleanup that page as well.

**Review by sdixon194 — DISMISSED** [review]

**Comment by jkguidaven** [comment]
> This works for me! Something I noted that might be good for a follow up, I still get the JITM that takes me to: https://causal-catfish.jurassic.ninja/wp-admin/admin.php?page=jetpack#/license/activation
> 
> Screen Shot 2022-12-12 at 10 01 15 AM
> 
> That license activation page seems different and doesn't let me pick existing licenses from a dropdown. Not sure if there are plans to cleanup that page as well.

Thanks @sdixon194 for reviewing this. I created a separate task for this to fix the JITM and cleanup the page. 👍

**Review by manzoorwanijk — APPROVED** [review]

**Comment by jkguidaven** [comment]
@sdixon194 sorry I lost your approval when I rebased my branch.. can you reapprove again so I can merge it? 🙏

**Review by sdixon194 — APPROVED** [review]

**Pull request merged**
